x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -151,15 +160,44 @@ protected ChannelHandler getClientChannelInitializer(DiscoveryNode node, Connect
 
     @Override
     protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChannel) {
-        return new InboundPipeline(
-            getStatsTracker(),
-            threadPool::relativeTimeInMillis,
-            isRemoteClusterServerChannel
-                ? new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings))
-                : new InboundDecoder(recycler),
-            new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
-            this::inboundMessage
-        );
+        if (false == isRemoteClusterServerChannel) {
+            return super.getInboundPipeline(false);
+        } else {
+            return new InboundPipeline(
+                getStatsTracker(),
+                threadPool::relativeTimeInMillis,
+                new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings), false),
+                new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
+                this::inboundMessage
+            ) {
+                @Override
+                protected void headerReceived(TcpChannel channel, Header header) {
+                    if (header.isHandshake() == false) {
+                        // eagerly (before buffering the full request) authenticate all request headers for this type of channel
+                        assert header.isRequest();
+                        // authn is mostly async, avoid buffering anymore data while authn is in progress
+                        boolean isNettyChannel = channel instanceof Netty4TcpChannel;
+                        if (isNettyChannel) {
+                            ((Netty4TcpChannel) channel).getNettyChannel().config().setAutoRead(false);
+                        }
+                        // this prevents thread-context changes to propagate beyond the validation, as netty worker threads are reused
+                        try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().newStoredContext()) {
+                            crossClusterAccessAuthenticationService.tryAuthenticate(
+                                header.getRequestHeaders(),
+                                ActionListener.wrap(aVoid -> {
+                                    // authn is successful -> NOOP (the complete request will be subsequently authn & authz & audited)
+                                    if (isNettyChannel) {
+                                        ((Netty4TcpChannel) channel).getNettyChannel().config().setAutoRead(true);
+                                    }
```

Comment:
@Tim-Brooks Is it OK to call this on a non-evenloop thread. I think it is, but can we rely on it? More generally, what do you think of disabling auto-read while authn is in progress, like implemented here?

Reply:
This is an interesting point. Yes functionally it should be fine here to continually disable autoread since propgating existing full messages does not rely on it. But it could be an useful optimisation for happy paths to avoid 2 atomic write per request.

Reply:
I get that we're OK to `setAutoRead(false)` before running authn and set it back to `true` when authn returns (either successfully or not), in this 8.10 quickfix implementation.